src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/CompiledRegexRunner.cs

```diff
@@ -1,23 +1,172 @@
+        {
+            this.quick = quick;
+
+            // Handle timeout argument
+            _timeout = -1; // (int)Regex.InfiniteMatchTimeout.TotalMilliseconds
+            bool ignoreTimeout = _ignoreTimeout = Regex.InfiniteMatchTimeout == timeout;
+            if (!ignoreTimeout)
+            {
+                // We are using Environment.TickCount and not Stopwatch for performance reasons.
+                // Environment.TickCount is an int that cycles. We intentionally let timeoutOccursAt
+                // overflow it will still stay ahead of Environment.TickCount for comparisons made
+                // in DoCheckTimeout().
+                Regex.ValidateMatchTimeout(timeout); // validate timeout as this could be called from user code due to being protected
+                _timeout = (int)(timeout.TotalMilliseconds + 0.5); // Round;
+                _timeoutOccursAt = Environment.TickCount + _timeout;
+                _timeoutChecksToSkip = TimeoutCheckFrequency;
+            }
+
+            // Configure the additional value to "bump" the position along each time we loop around
+            // to call FindFirstChar again, as well as the stopping position for the loop.  We generally
+            // bump by 1 and stop at textend, but if we're examining right-to-left, we instead bump
+            // by -1 and stop at textbeg.
+            int bump = 1, stoppos = textend;
+            if (regex.RightToLeft)
+            {
+                bump = -1;
+                stoppos = textbeg;
+            }
+
+            // Store runtextpos into field, as we may bump it in next check.  The remaining arguments
+            // are stored below once we're past the potential return in the next check.
+            runtextpos = textstart;
+
+            // If previous match was empty or failed, advance by one before matching.
+            if (prevlen == 0)
+            {
+                if (textstart == stoppos)
+                {
+                    runmatch = Match.Empty;
+                    return;
+                }
+
+                runtextpos += bump;
+            }
+
+            // Store remaining arguments into fields now that we're going to start the scan.
+            // These are referenced by the derived runner.
+            runregex = regex;
+            runtextstart = textstart;
+            runtextbeg = textbeg;
+            runtextend = textend;
+
+            // Main loop: FindFirstChar/Go + bump until the ending position.
+            bool initialized = false;
+            while (true)
+            {
+#if DEBUG
+                if (regex.IsDebug)
+                {
+                    Debug.WriteLine("");
+                    Debug.WriteLine($"Search range: from {runtextbeg} to {runtextend}");
+                    Debug.WriteLine($"Firstchar search starting at {runtextpos} stopping at {stoppos}");
+                }
+#endif
```

Comment:
I think we're reaching a point where all of this DEBUG stuff is no longer valuable.  We might want to strip it out.  Or at least, I don't use it. :)